### PR TITLE
Ptp master

### PR DESCRIPTION
**Pull request by bramtweedegolf**

### CODE REVIEW

File: statime/src/port.rs [hunk 1]
```diff
@@ -269,10 +285,148 @@ impl StateSlave {
...
+
+        if port.follow_up_seq_id == u16::MAX {
+            port.follow_up_seq_id = 0;
+        } else {
+            port.follow_up_seq_id += 1;
+        }
+
+        // TODO: This uses a Vec, is that allowed?
+        let follow_up_message_encode = follow_up_message.serialize_vec().unwrap();
+        port.tc_port.send(&follow_up_message_encode);
```
Comment (diondokter) [inline statime/src/port.rs:389]: Non tc port

File: statime/src/port.rs [hunk 2]
```diff
@@ -269,10 +285,148 @@ impl StateSlave {
...
+        // Send delay response
+        let delay_resp_message = MessageBuilder::new()
+            .sequence_id(port.delay_resp_seq_id)
+            .source_port_identity(port.identity)
+            .delay_resp_message(
+                Timestamp::from(timestamp),
+                message.header().source_port_identity(),
+            );
+
+        if port.delay_resp_seq_id == u16::MAX {
```
Comment (diondokter) [inline statime/src/port.rs:411]: Wrapping add

File: statime/src/port.rs [hunk 3]
```diff
@@ -269,10 +285,148 @@ impl StateSlave {
...
+            port.delay_resp_seq_id = 0;
+        } else {
+            port.delay_resp_seq_id += 1;
+        }
+
+        let delay_resp_encode = delay_resp_message.serialize_vec().unwrap();
+        port.tc_port.send(&delay_resp_encode);
```
Comment (diondokter) [inline statime/src/port.rs:418]: Non tc port?

File: statime/src/port.rs [hunk 4]
```diff
@@ -344,9 +602,92 @@ impl<NR: NetworkRuntime> Port<NR> {
...
+            announce_watch,
+            sync_watch,
         }
     }
 
+    pub fn handle_alarm(&mut self, id: W::WatchId, current_time: Instant) {
+        // When the announce timout expires, it means there
+        // have been no announce messages in a while, so we
+        // force a switch to the master state
+        if id == self.watch.id() {
```
Comment (diondokter) [inline statime/src/port.rs:615]: Change the name of this watch

File: statime/src/port.rs [hunk 5]
```diff
@@ -344,9 +602,92 @@ impl<NR: NetworkRuntime> Port<NR> {
...
+                self.portdata.port_config.announce_interval,
+            ));
+        }
+
+        // When the sync watch expires, send a sync message and restart
+        if id == self.sync_watch.id() {
+            self.send_sync_message(current_time);
+
+            // TODO: Is the follow up a config?
+            self.send_follow_up_message(current_time);
```
Comment (diondokter) [inline statime/src/port.rs:657]: Yes

**Review by diondokter — CHANGES_REQUESTED** [review]

File: statime/src/port.rs [hunk 6]
```diff
@@ -269,10 +285,148 @@ impl StateSlave {
...
+            .source_port_identity(port.identity)
+            .announce_message(
+                Timestamp::default(),             //origin_timestamp: Timestamp,
+                0,                                //current_utc_offset: u16,
+                self.priority_1,                  //grandmaster_priority_1: u8,
+                self.clock_quality,               //grandmaster_clock_quality: ClockQuality,
+                self.priority_2,                  //grandmaster_priority_2: u8,
+                self.clock_identity,              //grandmaster_identity: ClockIdentity,
+                0,                                //steps_removed: u16,
+                TimeSource::from_primitive(0xa0), //time_source: TimeSource,
```
[the remark this answers is not in the document]
Comment (bramtweedegolf) [reply statime/src/port.rs:330]: @diondokter Maybe I missed something in the code, but it seems like we don't establish a TimeSource anywhere at the moment. It should maybe be set by the Clock implementations? I added a TODO comment.

[the remark this answers is not in the document]
Comment (bramtweedegolf) [reply statime/src/port.rs]: Will implement this when the state-machine branch has been merged.

File: statime/src/port.rs (hunk 5 above)
Comment (bramtweedegolf) [reply statime/src/port.rs:632]: Will implement after the state-machine branch has been merged.

File: statime/src/port.rs [hunk 7]
```diff
@@ -295,26 +422,129 @@ impl State {
...
                 State::Listening => {
                     *self = State::Slave(StateSlave {
                         remote_master: announce_message.header().source_port_identity(),
                         ..Default::default()
-                    })
+                    });
+
+                    // Restart announce receipt timeout timer
+                    announce_timeout_watch
+                        .set_alarm(Duration::from_log_interval(announce_receipt_timeout));
```
Comment (Tortoaster) [inline statime/src/port.rs:449]: `announce_receipt_timeout` is zelf geen log interval, maar het aantal keren dat `announce_interval` herhaald wordt voor een `ANNOUNCE_RECEIPT_TIMEOUT_EXPIRES` event (7.7.3.1).

File: statime/src/port.rs [hunk 8]
```diff
@@ -295,26 +422,129 @@ impl State {
...
+                        ..Default::default()
+                    });
+
+                    // Stop MASTER watches
+                    announce_watch.clear();
+                    sync_watch.clear();
+
+                    // Restart announce receipt timeout timer
+                    announce_timeout_watch
+                        .set_alarm(Duration::from_log_interval(announce_receipt_timeout));
```
Comment (Tortoaster) [inline statime/src/port.rs:477]: ^

File: statime/src/port.rs [hunk 9]
```diff
@@ -295,26 +422,129 @@ impl State {
...
+                _ => {
+                    *self = State::Listening;
+
+                    // Stop MASTER watches
+                    announce_watch.clear();
+                    sync_watch.clear();
+
+                    // Restart announce receipt timeout timer
+                    announce_timeout_watch
+                        .set_alarm(Duration::from_log_interval(announce_receipt_timeout));
```
Comment (Tortoaster) [inline statime/src/port.rs:538]: ^

**Review by Tortoaster — CHANGES_REQUESTED** [review]

File: statime/src/ptp_instance.rs [hunk 10]
```diff
@@ -40,18 +40,31 @@ impl<NR: NetworkRuntime, C: Clock, F: Filter> PtpInstance<NR, C, F> {
             config.port_config.log_announce_interval,
         ));
 
+        // Set the announce receipt timeout
+        // TODO: what to do when we have multiple ports?
+        let mut announce_timeout_watch = clock.get_watch();
+        announce_timeout_watch.set_alarm(Duration::from_log_interval(
+            config.port_config.announce_receipt_timeout,
```
Comment (Tortoaster) [inline statime/src/ptp_instance.rs:47]: ^

**Review by Tortoaster — CHANGES_REQUESTED** [review]

File: statime/src/port.rs [hunk 11]
```diff
@@ -368,10 +685,17 @@ impl<NR: NetworkRuntime> Port<NR> {
...
-                .register_announce_message(&announce, current_time.into()),
+            Message::Announce(announce) => {
+                self.portdata
+                    .bmca
+                    .register_announce_message(&announce, current_time.into());
+
+                // When an announce message is received, restart announce receipt timeout timer
+                self.announce_timeout_watch
+                    .set_alarm(Duration::from_log_interval(
+                        self.portdata.port_config.announce_receipt_timeout,
```
Comment (Tortoaster) [inline statime/src/port.rs:696]: ^

**Review by Tortoaster — CHANGES_REQUESTED** [review]

File: statime/src/port.rs (hunk 7 above)
Comment (bramtweedegolf) [reply statime/src/port.rs]: Well spotted!

**Pull request closed without merging**